pytorch_lightning/trainer/supporters.py

```diff
@@ -118,27 +122,278 @@ def mean(self):
         return self.total / self.num_values
 
 
+class LightningBatchSamplerWrapper:
+    """
+    This class wraps user batch sampler, so we can extract
+    the batch_indices for tracking each sample.
+    """
+
+    SKIP_KEYS = ['sampler', 'batch_sampler', 'dataset_kind']
+    SKIP_VALID_KEYS = ['args', 'kwargs', 'self']
+
+    def __init__(self, batch_sampler):
+        self.batch_sampler = batch_sampler
+        self.batch_indices = None
+
+    def __iter__(self):
+        for batch_indices in self.batch_sampler:
+            self.batch_indices = batch_indices
+            yield batch_indices
+
+    @staticmethod
+    def recreate_dataloader(dataloader: DataLoader) -> DataLoader:
+        """
+        This function will wrap the user batch_sampler to track the returned batch indices
+        """
+        if not isinstance(dataloader, DataLoader):
+            raise MisconfigurationException('Autoid only works with torch dataloaders or derived classes!')
+
+        elif isinstance(dataloader.dataset, IterableDataset):
+            return dataloader
+
+        params = {k: v for k, v in vars(dataloader).items() if not k.startswith("_")}
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
+        contains_dataset = True
+
+        if type(dataloader) is not DataLoader:
+            contains_dataset = "dataset" in valid_kwargs
+            valid_kwargs.update(inspect.signature(DataLoader.__init__).parameters)
+
+        dl_args = {
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in LightningBatchSamplerWrapper.SKIP_KEYS
+        }
+
+        multiprocessing_context = dataloader.multiprocessing_context
+
+        # override parameters to enable batch_sampler injection
+        dl_args.update({
+            "batch_size": 1,
+            "sampler": None,
+            "shuffle": None,
+            "batch_sampler": LightningBatchSamplerWrapper(dataloader.batch_sampler),
+            "drop_last": False,
+            "multiprocessing_context": multiprocessing_context,
+        })
+
+        missing_kwargs = valid_kwargs.difference(LightningBatchSamplerWrapper.SKIP_VALID_KEYS).difference(dl_args)
+        if missing_kwargs:
+            dataloader_cls_name = dataloader.__class__.__name__
+            rank_zero_warn(
+                f"Trying to replace to wrap your BatchSampler in your {dataloader_cls_name} dataloader."
+                "This would fail as your DataLoader doesn't expose as attributes all its __init__ parameters. "
+                f"Missing attributes are {missing_kwargs} \n"
+                "HINT: use Trainer(enable_predict_auto_id=False) and provide your own id."
```

Comment:
i don't know what this does....
can you rename the flag to make it clear? or add docstrings or a much clearer error message?